### PR TITLE
Allow users to log into cashbook

### PR DESCRIPTION
**Pull request by marcofucci**
This allows a valid user to log into the cashbook app using oauth2 against the API app.

Custom login, logout and authentication logic is required as the standard django behaviour
relies on database tables.
Unfortunately, I had to borrow some django code and change it accordingly in order to support our authentication method.

Use the mock python library for mocking the API connection so that unit tests can run
without dependencies.

### CODE REVIEW

**Comment by marcofucci** [comment]
@r4vi @SteveMarshall do you want to have a look?

File: mtp_cashbook/apps/mtp_auth/exceptions.py [hunk 1]
```diff
@@ -0,0 +1,6 @@
+class ConnectionError(Exception):
+    """
+    Exception that should be raised in case of connection
+    problems with the api server.
+    """
+    pass
```
Comment (SteveMarshall) [inline mtp_cashbook/apps/mtp_auth/exceptions.py]: Do we need our own `ConnectionError` class? Isn't this a little confusing, given that [Python 3 provides a `ConnectionError`](https://docs.python.org/3/library/exceptions.html#ConnectionError)? Should we just use that (or make this a subclass… or just pass out whatever the underlying error is)?

File: mtp_cashbook/apps/mtp_auth/forms.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    """
+    Authentication form used for authenticating users during the login process.
+    """
+    username = forms.CharField(label=_("Username"), max_length=254)
+    password = forms.CharField(label=_("Password"), widget=forms.PasswordInput)
+
+    error_messages = {
+        'invalid_login': _("Please enter a correct username and password. "
+                           "Note that both fields may be case-sensitive."),
+        'connection_error': _("The API Server seems down, please try again later."),
```
Comment (SteveMarshall) [inline mtp_cashbook/apps/mtp_auth/forms.py]: I'd be tempted to explicitly make these translation keywords, rather than the literal english text we expect them to be. That way, when @leighmoney (or whoever) wants to improve the copy, we don't have to dig around in the code to make that change.

Comment (SteveMarshall) [inline mtp_cashbook/apps/mtp_auth/forms.py]: That would also mean that we can get rid of the `error_messages` hash here and just use `_()` directly through the code. I'm not sure if I prefer that or not, mind you.

File: mtp_cashbook/apps/mtp_auth/forms.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    def __init__(self, request=None, *args, **kwargs):
+        self.request = request
+        self.user_cache = None
```
Comment (SteveMarshall) [inline mtp_cashbook/apps/mtp_auth/forms.py]: Why `user_cache`? Unless I'm missing something, this feels more like `current_user`?

File: mtp_cashbook/apps/mtp_auth/forms.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        super(AuthenticationForm, self).__init__(*args, **kwargs)
+
+    def clean(self):
```
Comment (SteveMarshall) [inline mtp_cashbook/apps/mtp_auth/forms.py]: This doesn't feel like `clean` to me: it seems like this it's getting the cleaned inputs and then doing the actual login… Oh, wait, is this a holdover of this being a subclass of `django.forms.Form`? If so, it might be nice to break this out more clearly into a `login` method that the `clean` method calls, perhaps?

File: mtp_cashbook/apps/mtp_auth/forms.py (hunk 3 above)
Comment (SteveMarshall) [inline mtp_cashbook/apps/mtp_auth/forms.py]: This might also reduce the need for the comment about authenticate returning `None` below.

File: mtp_cashbook/apps/mtp_auth/exceptions.py (hunk 1 above)
Comment (marcofucci) [inline mtp_cashbook/apps/mtp_auth/exceptions.py]: Yes I knew that you might have not liked this.
It's not confusing for me as I wanted to make explicit that you need to catch **this** exception and understand what it means for you as dev when you do it.
If it's confusing for you, I might just use the requests one or the python3 one, doesn't really change anything for me really.

File: mtp_cashbook/apps/mtp_auth/middleware.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+
+def get_user(request):
+    """
+    Returns a cached copy of the user if it exists or calls `auth_get_user`
+    otherwise.
+    """
+    if not hasattr(request, '_cached_user'):
+        request._cached_user = auth_get_user(request)
+    return request._cached_user
```
Comment (SteveMarshall) [inline mtp_cashbook/apps/mtp_auth/middleware.py]: Again: `_cached_user` might be clearer as `_current_user` unless there's some nuance I'm missing?

File: mtp_cashbook/apps/mtp_auth/models.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
+class MtpUser(object):
+    """
+    Authenticated user, similar to the Django one.
+
+    The built-in Django `AbstractBaseUser` sadly depends on a few tables and
+    cannot be used without a datbase so we had to create a custom one.
```
Comment (SteveMarshall) [inline mtp_cashbook/apps/mtp_auth/models.py]: This stuff would make perfect material for a commit message rather than a comment, if this were done in multiple discreet commits :wink:

File: mtp_cashbook/apps/mtp_auth/forms.py (hunk 2 above)
Comment (marcofucci) [inline mtp_cashbook/apps/mtp_auth/forms.py]: Those are just keys as any other keys, you will be able to change them in the translation files if you want although I doubt we need translation files.

This is standard django convention, explicit english translation keys and error_messages hash.
I suggest we shouldn't change it as it would be too confusing for devs.

File: mtp_cashbook/apps/mtp_auth/forms.py (hunk 3 above)
Comment (marcofucci) [inline mtp_cashbook/apps/mtp_auth/forms.py]: Wanted to keep it the same as django which uses this name, I don't see any problems really

File: mtp_cashbook/templates/base.html [hunk 7]
```diff
@@ -21,7 +21,7 @@
   {% endblock content %}
 
   {% if DEBUG %}
-  
+
   <script src="{% static "javascripts/app.js" %}"></script>
   <script src="//localhost:35729/livereload.js"></script>
```
Comment (r4vi) [inline mtp_cashbook/templates/base.html]: not part of this PR but this needs to probably be deleted

**Comment by r4vi** [comment]
Looks good to me :+1:

File: mtp_cashbook/apps/mtp_auth/views.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+        request.GET.get(redirect_field_name, '')
+    )
+
+    if request.method == "POST":
+        form = authentication_form(request, data=request.POST)
+        if form.is_valid():
+
+            # Ensure the user-originating redirection url is safe.
+            if not is_safe_url(url=redirect_to, host=request.get_host()):
+                redirect_to = resolve_url(settings.LOGIN_REDIRECT_URL)
```
Comment (SteveMarshall) [inline mtp_cashbook/apps/mtp_auth/views.py]: :+1:

File: mtp_cashbook/apps/mtp_auth/forms.py (hunk 4 above)
Comment (marcofucci) [inline mtp_cashbook/apps/mtp_auth/forms.py]: yes, it's the django `clean` method where this logic should be.

I think the code is clear enough for django devs. If you have a different `login` method, it wouldn't be clear that you can't call it before calling clean so you would expect to call `login` and just work.

File: mtp_cashbook/templates/mtp_auth/login.html [hunk 9]
```diff
@@ -0,0 +1,39 @@
+{% extends "base.html" %}
+
+{% block content %}
+{% if form.errors and not form.non_field_errors %}
+<p class="errornote">
+{% if form.errors.items|length == 1 %}Please correct the error below.{% else %}Please correct the errors below{% endif %}
```
Comment (SteveMarshall) [inline mtp_cashbook/templates/mtp_auth/login.html]: Can we wrap this for translation?

File: mtp_cashbook/apps/mtp_auth/forms.py (hunk 2 above)
Comment (SteveMarshall) [inline mtp_cashbook/apps/mtp_auth/forms.py]: This is the standard in Django? Huh. Well, let's stick with this then, though the English-as-default thing always makes me uncomfortable.

File: mtp_cashbook/apps/mtp_auth/forms.py (hunk 3 above)
Comment (SteveMarshall) [inline mtp_cashbook/apps/mtp_auth/forms.py]: If this is following Django convention, then OK!

File: mtp_cashbook/apps/mtp_auth/forms.py (hunk 2 above)
Comment (marcofucci) [inline mtp_cashbook/apps/mtp_auth/forms.py]: well, it's not english as default but more "A string like this" is better than A_STRING_LIKE_THIS.
Does it make more sense?

Comment (SteveMarshall) [inline mtp_cashbook/apps/mtp_auth/forms.py]: I'm not sure I buy that argument, but that's one for the pub :beers: Leave it as it is, for the reasons you say.

File: mtp_cashbook/apps/mtp_auth/middleware.py (hunk 5 above)
Comment (marcofucci) [inline mtp_cashbook/apps/mtp_auth/middleware.py]: Again, that's how django called that variable so I don't want to change it unless it's really necessary.

File: mtp_cashbook/templates/mtp_auth/login.html (hunk 9 above)
Comment (marcofucci) [inline mtp_cashbook/templates/mtp_auth/login.html]: Do we need translation?

File: mtp_cashbook/apps/mtp_auth/forms.py (hunk 4 above)
Comment (SteveMarshall) [inline mtp_cashbook/apps/mtp_auth/forms.py]: Having had a poke around Django I see that, yeah, you're right. I'll shut up :stuck_out_tongue:

File: mtp_cashbook/templates/mtp_auth/login.html (hunk 9 above)
Comment (SteveMarshall) [inline mtp_cashbook/templates/mtp_auth/login.html]: Well, we're doing it elsewhere, and we'll need to support Welsh at some point so if it's easy-ish, we should do it.

File: mtp_cashbook/apps/mtp_auth/middleware.py (hunk 5 above)
Comment (SteveMarshall) [inline mtp_cashbook/apps/mtp_auth/middleware.py]: Yup, fine. To make a point: we wouldn't be having this discussion if this were in a discrete commit explaining why we're rolling our own and that we're basically mirroring the Django core auth with small changes to support oauth.

File: mtp_cashbook/apps/mtp_auth/exceptions.py (hunk 1 above)
Comment (SteveMarshall) [inline mtp_cashbook/apps/mtp_auth/exceptions.py]: No, that's fine by me.
